search-sentiments-map/routes/search.js

```diff
@@ -28,20 +28,135 @@ global.Headers = fetch.Headers;
 const STALE_SEARCH_RESULT_THRESHOLD_7_DAYS_MS = 7 * 24 * 60 * 60000;
 
 /** 
- * Renders a JSON array of the top search results for all countries with API data
- * obtained every 12 hours for the specified topic.
+ * Renders a JSON array of the top search results for all countries with API
+ * data obtained every 12 hours for the specified topic.
  */
 router.get('/:topic', (req, res) => {
   let topic = req.params.topic;
   retrieveSearchResultFromDatastore(topic).then(customSearchTopicJsonArray => {
-    // TODO(ntarn): Remove console.log statements when finished debugging.
-    console.log('ntarn debug: In the router.get method'); 
-    console.log(customSearchTopicJsonArray);
     res.setHeader('Content-Type', 'application/json');
     res.send(customSearchTopicJsonArray);
   });
 });
 
+/** 
+ * Renders a JSON array of the top search results for requested countries with
+ * API data from within the last 12 hours for the specified topic.
+ */
+router.get('/:topic/:countries', (req, res) => {
+  let topic = req.params.topic;
+  let countries = JSON.parse(req.params.countries);
+  retrieveUserSearchResultFromDatastore(topic, countries).then(userSearchTopicJsonArray => {
+    res.setHeader('Content-Type', 'application/json');
+    res.send(userSearchTopicJsonArray);
+  });
+});
+
+/** 
+ * Retrieves search result data for specified countries for a given topic.
+ *   - Obtains and formats search results from Custom Search API if the data
+ *         is not currently stored.
+ *   - Saves any new search results to the datastore.
+ *   - Retrieves any relevant existing search results from datastore.
+ *   - Returns all requested search results on the topic from the countries.
+ * @param {string} topic Search topic to get data for.
+ * @param {Array} countries Countries to get search topic data for.
+ * @return {Object} JSON-formatted array of search results from requested
+ *     countries.
+ */
+async function retrieveUserSearchResultFromDatastore(topic, countries) {
+  // Request latest entity with a topic matching the given topic.
+  const query = datastore.createQuery('CustomSearchTopic').order('timestamp', {
+    descending: true,
+  }).filter('lowercaseTopic', topic.toLowerCase()).limit(1);
+  const [customSearchTopic] = await datastore.runQuery(query);
+  // TODO(carmenbenitez): Handle capitalization differences. Change to last 12 hours
+
+  let countriesDataToReturn = [];
+  let timestamp;
+  if (customSearchTopic.length !== 0 &&
+      Date.now() - customSearchTopic[0].timestamp < 7 * 24 * 60 * 60000) {
+      timestamp = customSearchTopic[0].timestamp;
+      let countriesToAddDataFor = [];
+
+      // Determine whether a country has existing data or data needs to be
+      // retrieved from Custom Search API for this country on this topic.
```

Comment:
Wouldn't all countries have data for the same set of topics? So all of them need to retrieve data from API or not. Why would you need to treat it on a country-by-country basis?

Reply:
We are only considering the most recently retrieve data set right?

Reply:
Because what happens is if a user searches a new topic, for example "coronavirus", and they select their countries to view results for, for example Australia and Canada, we create a custom search entity for coronavirus with only data for Australia and Canada.

If later that day, another user searches Coronavirus, and this time they select the US and Canada, the function would look and determine that we already have data for Canada(and therefore no API calls are needed), but we need data for the US. It would then retrieve that data just for the US and add that data to the entity.

Reply:
Ah gotcha.